### PR TITLE
The mc now keeps track of how many times a subsystem has slept in fire()

### PR DESCRIPTION
**Pull request by MrsTonedOne**
## About The Pull Request

The mc now keeps track of how many times a subsystem has slept in fire()

## Why It's Good For The Game

The mc now keeps track of how many times a subsystem has slept in fire()

## Changelog

The mc now keeps track of how many times a subsystem has slept in fire()

### CODE REVIEW

File: code/controllers/subsystem.dm [hunk 1]
```diff
@@ -65,6 +65,9 @@
 
 	/// Tracks the current execution state of the subsystem. Used to handle subsystems that sleep in fire so the mc doesn't run them again while they are sleeping
 	var/state = SS_IDLE
+	
+	/// Tracks how many times a subsystem has ever slept in fire().
+	var/slept_count = 0
```
Comment (LemonInTheDark) [inline code/controllers/subsystem.dm:70]: can we log it? would be good to know this info as a metric to improve on, where it's not intended at least

**Comment by AffectedArc07** [comment]
The mc now keeps track of how many times a subsystem has slept in fire()

**Comment by TrashDoxx** [comment]
The mc now keeps track of how many times a subsystem has slept in fire()

**Comment by Mothblocks** [comment]
I was under the impression sleeping in fire() was fine, and the documentation doesn't seem to suggest it is wrong. Your PR doesn't description really provide any new hints so I'm going to assume this is something I needed to be in Discord for. Is this behavior that we do not want, ever? If so then we can probably have a PR after this one flipping on `SHOULD_NOT_SLEEP(TRUE)`

**Comment by AffectedArc07** [comment]
> I was under the impression sleeping in fire() was fine, and the documentation doesn't seem to suggest it is wrong. Your PR doesn't description really provide any new hints so I'm going to assume this is something I needed to be in Discord for. Is this behavior that we do not want, ever? If so then we can probably have a PR after this one flipping on `SHOULD_NOT_SLEEP(TRUE)` and pruning consumers

This is ill advised due to some subsystems executing DB queries or running emotes that can sleep from SSmobs or similar.

**Comment by Mothblocks** [comment]
If there are legitimate cases then "warning" is a bit odd? I was just expecting a blackbox

**Comment by AffectedArc07** [comment]
> If there are legitimate cases then "warning" is a bit odd? I was just expecting a blackbox

The MC /shouldnt/ be sleeping, but when I made fire unsleepable on our code we had 600+ errors to fix, this is likely to be even worse with deferred verbs and client/Topic, since things like adding notes that make sleeping DB queries will now trip this.

**Comment by LemonInTheDark** [comment]
image

**Comment by MrStonedOne** [comment]
> I was under the impression sleeping in fire() was fine, 

it is

> and the documentation doesn't seem to suggest it is wrong. 

its not ideal since in most cases its things that fire calls that should be sleeping.

> Your PR doesn't description really provide any new hints so I'm going to assume this is something I needed to be in Discord for. 

correct.

> Is this behavior that we do not want, ever? If so then we can probably have a PR after this one flipping on `SHOULD_NOT_SLEEP(TRUE)` and pruning consumers

na, the mc can handle it. its a case where a lot of cases of sleep is likely to be an issue for the subsystem, but not the mc, so the mc should track how often it happens so subsystem coders notice and think about if thats what they want.

Adding the 'warning' was done at request, the primary goal is to be informative.

**Comment by Mothblocks** [comment]
Works for me but I'm admittedly a little surprised that you of all people don't mind opening pull requests without putting the context in the PR itself (Lemon's screenshot makes this PR make significantly more sense to me)

**Pull request merged**